File: Weekly Contest 366/100085. Minimum Processing Time.py
import heapq
from typing import List

class Solution:
    def minProcessingTime(self, processorTime: List[int], tasks: List[int]) -> int:
        n = len(processorTime)
        processorTime.sort()
        tasks.sort(reverse=True)  # Sort tasks in decreasing order of execution time
        heap = [(processorTime[i], i) for i in range(n)]  # (available time, processor index)
        total_time = 0
        
        for task_time in tasks:
            if not heap:
                total_time += task_time
                continue
            avail_time, processor_idx = heapq.heappop(heap)
            avail_time += task_time
            heapq.heappush(heap, (avail_time, processor_idx))
        
        while heap:
            avail_time, processor_idx = heapq.heappop(heap)
            total_time = max(total_time, avail_time - processorTime[processor_idx])
        
        return total_time

# Example usage:
solution = Solution()
processorTime1 = [8,10]
tasks1 = [2,2,3,1,8,7,4,5]
print(solution.minProcessingTime(processorTime1, tasks1))  # Output: 16

File: Weekly Contest 366/100085. Minimum Processing Time.cs


public class Solution
{
    public int MinProcessingTime(int[] processorTime, int[] tasks)
    {
        int n = processorTime.Length;
        int m = tasks.Length;

        Array.Sort(processorTime, (a, b) => b - a); // Sort processors in descending order.

        long left = 0;
        long right = (long)processorTime[0] * m; // Maximum possible time

        while (left < right)
        {
            long mid = left + (right - left) / 2;

            if (IsFeasible(processorTime, tasks, mid))
            {
                right = mid;
            }
            else
            {
                left = mid + 1;
            }
        }

        return (int)left;
    }

    private bool IsFeasible(int[] processorTime, int[] tasks, long maxTime)
    {
        int n = processorTime.Length;
        int m = tasks.Length;
        int taskIndex = m - 1;

        for (int i = 0; i < n; i++)
        {
            long timeLeft = maxTime / processorTime[i];

            while (taskIndex >= 0 && tasks[taskIndex] <= timeLeft)
            {
                timeLeft -= tasks[taskIndex];
                taskIndex--;
            }
        }

        return taskIndex < 0;
    }
}
